pkg/chartmuseum/server/multitenant/cache.go

```diff
@@ -0,0 +1,313 @@
+	}
+	log(cm_logger.DebugLevel, "Updating chart in index",
+		"repo", repo,
+		"name", chartVersion.Name,
+		"version", chartVersion.Version,
+	)
+	index.UpdateEntry(chartVersion)
+	return nil
+}
+
+func (server *MultiTenantServer) addIndexObjectsAsync(log cm_logger.LoggingFn, repo string, index *cm_repo.Index, objects []cm_storage.Object) error {
+	numObjects := len(objects)
+	if numObjects == 0 {
+		return nil
+	}
+
+	log(cm_logger.DebugLevel, "Loading charts packages from storage (this could take awhile)",
+		"repo", repo,
+		"total", numObjects,
+	)
+
+	type cvResult struct {
+		cv  *helm_repo.ChartVersion
+		err error
+	}
+
+	cvChan := make(chan cvResult)
+
+	// Provide a mechanism to short-circuit object downloads in case of error
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for _, object := range objects {
+		go func(o cm_storage.Object) {
+			if server.IndexLimit != 0 {
+				// Limit parallelism to the index-limit parameter value
+				// if there are more than IndexLimit concurrent fetches, this send will block
+				server.Limiter <- struct{}{}
+				// once work is over, read one Limiter channel item to allow other workers to continue
+				defer func() { <-server.Limiter }()
+			}
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				chartVersion, err := server.getObjectChartVersion(repo, o, true)
+				if err != nil {
+					err = server.checkInvalidChartPackageError(log, repo, o, err, "added")
+					if err != nil {
+						cancel()
+					}
+				}
+				cvChan <- cvResult{chartVersion, err}
+			}
+		}(object)
+	}
+
+	for validCount := 0; validCount < numObjects; validCount++ {
+		cvRes := <-cvChan
+		if cvRes.err != nil {
+			return cvRes.err
+		}
+		if cvRes.cv == nil {
+			continue
+		}
+		log(cm_logger.DebugLevel, "Adding chart to index",
+			"repo", repo,
+			"name", cvRes.cv.Name,
+			"version", cvRes.cv.Version,
+		)
+		index.AddEntry(cvRes.cv)
+	}
+
+	return nil
+}
+
+func (server *MultiTenantServer) getObjectChartVersion(repo string, object cm_storage.Object, load bool) (*helm_repo.ChartVersion, error) {
+	op := object.Path
+	if load {
+		var err error
+		objectPath := fmt.Sprintf("%s/%s", repo, op)
+		object, err = server.StorageBackend.GetObject(objectPath)
+		if err != nil {
+			return nil, err
+		}
+		if len(object.Content) == 0 {
+			return nil, cm_repo.ErrorInvalidChartPackage
+		}
+	}
+	return cm_repo.ChartVersionFromStorageObject(object)
+}
+
+func (server *MultiTenantServer) checkInvalidChartPackageError(log cm_logger.LoggingFn, repo string, object cm_storage.Object, err error, action string) error {
+	if err == cm_repo.ErrorInvalidChartPackage {
+		log(cm_logger.WarnLevel, "Invalid package in storage",
+			"repo", repo,
+			"action", action,
+			"package", object.Path,
+		)
+		return nil
+	}
+	return err
+}
+
+func (server *MultiTenantServer) initCachedIndexFile(log cm_logger.LoggingFn, repo string) {
+	if _, ok := server.IndexCache[repo]; !ok {
```

Comment:
I think there is a race here. IndexCache could be evaluated empty for a given repo while another thread is initializing the repo cache.  Why not move the Lock() on function entry ? (then you could defer right away also)